cpp/src/arrow/engine/substrait/serde.cc

```diff
@@ -236,14 +236,13 @@ ARROW_ENGINE_EXPORT Result<PlanInfo> DeserializePlan(
 
 namespace {
 
-Result<std::shared_ptr<acero::ExecPlan>> MakeSingleDeclarationPlan(
-    std::vector<acero::Declaration> declarations) {
+Status MakeSingleDeclarationPlan(acero::ExecPlan* plan,
```

Comment:
This allows one to pass a preconfigured plan instead of getting a default-configured plan. The tester uses this to pass a plan with a non-default executor `ctx_for_ordered`.

Reply:
I think you can use https://github.com/apache/arrow/blob/c3acc91c3e21f225ee7cc3cb3f489f0153365f5a/cpp/src/arrow/acero/exec_plan.h#L566

Instead and can avoid changing the serde files

Reply:
i.e., instead of de serializing into a plan, you can de serializing into a declaration and use single-threaded or multi-threaded execution depending on ordered/non-ordered.

Reply:
Done. It turned out not easy to use `DeclarationToTable` here because of the extra setup needed for linking the input table and the consumer, but I did separate out the deserialization and obtained simpler code.